### PR TITLE
big endian

### LINKED ISSUE #15: Crate will not work correctly on big-endian architecture
Usage of `mem::transmute` will cause your code to produce incorrect results on big-endiann architectures.

You can adopt the same [approach](https://github.com/RustCrypto/hashes/blob/master/sha3/src/lib.rs#L80) I've used in the sha3 crate.

### PR DESCRIPTION
**Pull request by debris**
closes #15 
closes #35

### CODE REVIEW

**Comment by debris** [comment]
@tomusdrw please review :)

File: src/lib.rs [hunk 1]
```diff
@@ -131,26 +131,70 @@ pub fn keccakf(a: &mut [u64; PLEN]) {
...
-        unsafe {
-            *dst_ptr ^= *src_ptr;
-            src_ptr = src_ptr.offset(1);
-            dst_ptr = dst_ptr.offset(1);
+    #[cfg(target_endian = "big")]
+    #[inline]
+    fn execute<F: FnOnce(&mut [u8])>(&mut self, offset: usize, len: usize, f: F) {
+        fn swap_endianess(buffer: &mut [u64]) {
+            for item in buffer {
+                let i: &mut [u8; 8] = unsafe { ::core::mem::transmute(item) };
```
Comment (tomusdrw) [inline src/lib.rs]: Can't we use `item.to_be()`? Or is this code here in place to avoid copying `u64`?

File: src/lib.rs [hunk 2]
```diff
@@ -234,7 +278,7 @@ impl_global_alias!(sha3_512, 512);
 impl Keccak {
     pub fn new(rate: usize, delim: u8) -> Keccak {
```
Comment (tomusdrw) [inline src/lib.rs]: Should probably panic early in case `rate == 0` to avoid overflows later on when we subtract 1

File: src/lib.rs [hunk 3]
```diff
@@ -131,26 +131,70 @@ pub fn keccakf(a: &mut [u64; PLEN]) {
...
-    let mut dst_ptr = dst.as_mut_ptr();
-    let mut src_ptr = src.as_ptr();
-    for _ in 0..len {
-        unsafe {
-            *dst_ptr ^= *src_ptr;
-            src_ptr = src_ptr.offset(1);
-            dst_ptr = dst_ptr.offset(1);
+    #[cfg(target_endian = "big")]
+    #[inline]
+    fn execute<F: FnOnce(&mut [u8])>(&mut self, offset: usize, len: usize, f: F) {
```
Comment (tomusdrw) [inline src/lib.rs]: It seems a bit wasteful to actually convert to and back again, wouldn't it be better to modify `f`s in a endianess-agnostic way?

File: src/lib.rs (hunk 1 above)
Comment (debris) [reply src/lib.rs]: `to_be` is no-op on big endian platforms (and this function is executed on only on big-endian platform), but I guess I could also use `swap_bytes`. I don't think copying matters a lot here. I'll benchmark it

File: src/lib.rs (hunk 3 above)
Comment (debris) [reply src/lib.rs]: I agree and that's what I've tried in a first place. But then I realised that the number of operations (multiplication and rem) to calculate the correct indexes of bytes is even more expensive and simply swapping the endianess of those we use

**Comment by debris** [comment]
> Looks sane, but do we have any benchmarks to prove the performance?

Thanks for the review! Unfortunately we don't have. For now I focused only on making it work and not decreasing performance on little-endian platforms

**Pull request merged**